src/NServiceBus.RavenDB/Outbox/RavenDbOutboxStorage.cs

```diff
@@ -0,0 +1,110 @@
+﻿namespace NServiceBus.RavenDB.Outbox
+{
+    using System;
+    using System.Configuration;
+    using System.Threading;
+    using NServiceBus.Features;
+    using NServiceBus.Outbox.RavenDB;
+    using NServiceBus.RavenDB.Internal;
+    using Raven.Client;
+
+    class RavenDbOutboxStorage : Feature
+    {
+        public RavenDbOutboxStorage()
+        {
+            DependsOn<Outbox>();
+            DependsOn<SharedDocumentStore>();
+            RegisterStartupTask<OutboxCleaner>();
+        }
+
+        protected override void Setup(FeatureConfigurationContext context)
+        {
+            var store =
+                // Trying pulling a shared DocumentStore set by the user or other Feature
+                context.Settings.GetOrDefault<IDocumentStore>(RavenDbSettingsExtensions.DocumentStoreSettingsKey) ?? SharedDocumentStore.Get(context.Settings);
+
+            if (store == null)
+            {
+                throw new Exception("RavenDB is configured as persistence for Outbox and no DocumentStore instance found");
+            }
+
+            ConnectionVerifier.VerifyConnectionToRavenDBServer(store);
+
+            Helpers.SafelyCreateIndex(store, new OutboxRecordsIndex());
+
+            context.Container.ConfigureComponent<Installer>(DependencyLifecycle.InstancePerCall)
+                .ConfigureProperty(c => c.StoreToInstall, store);
+
+            context.Container.ConfigureComponent<OutboxPersister>(DependencyLifecycle.InstancePerCall)
+                .ConfigureProperty(x => x.DocumentStore, store);
+
+            context.Container.ConfigureComponent<OutboxRecordsCleaner>(DependencyLifecycle.InstancePerCall)
+                .ConfigureProperty(x => x.DocumentStore, store);
+        }
+
+        class OutboxCleaner : FeatureStartupTask, IDisposable
+        {
+            Timer cleanupTimer;
+            TimeSpan frequencyToRunDeduplicationDataCleanup;
+            TimeSpan timeToKeepDeduplicationData;
+            public OutboxRecordsCleaner Cleaner { get; set; }
+
+            public void Dispose()
+            {
+                if (cleanupTimer != null)
+                {
+                    cleanupTimer.Dispose();
+                    cleanupTimer = null;
+                }
+            }
+
+            protected override void OnStart()
+            {
+                var configValue = ConfigurationManager.AppSettings.Get("NServiceBus/Outbox/RavenDB/TimeToKeepDeduplicationData");
```

Comment:
Yes, that one is already released so we need to think about backwards
compat though (and deprecating a app.config api is a PITA)

On Mon, Apr 13, 2015 at 2:26 PM, Daniel Marbach <notifications@github.com>
wrote:

> In src/NServiceBus.RavenDB/Outbox/RavenDbOutboxStorage.cs
> <https://github.com/Particular/NServiceBus.RavenDB/pull/43#discussion_r28233523>
> :
>
> > +            TimeSpan frequencyToRunDeduplicationDataCleanup;
> > +            TimeSpan timeToKeepDeduplicationData;
> > +            public OutboxRecordsCleaner Cleaner { get; set; }
> > +
> > +            public void Dispose()
> > +            {
> > +                if (cleanupTimer != null)
> > +                {
> > +                    cleanupTimer.Dispose();
> > +                    cleanupTimer = null;
> > +                }
> > +            }
> > +
> > +            protected override void OnStart()
> > +            {
> > +                var configValue = ConfigurationManager.AppSettings.Get("NServiceBus/Outbox/RavenDB/TimeToKeepDeduplicationData");
>
> So we move RavenDB first and later NH as well?
> https://github.com/Particular/NServiceBus.NHibernate/blob/develop/src/NServiceBus.NHibernate/Outbox/NHibernateOutboxStorage.cs#L64
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Particular/NServiceBus.RavenDB/pull/43/files#r28233523>
> .
>